kms-message/src/kms_kmip_request.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kms_kmip_reader_writer_private.h"
+#include "kms_message/kms_kmip_request.h"
+#include "kms_message_private.h"
+
+#include <inttypes.h>
+#include <stdint.h>
+
+static void
+copy_writer_buffer (kms_request_t *req, kmip_writer_t *writer) {
+   const uint8_t *buf;
+   size_t buflen;
+
+   buf = kmip_writer_get_buffer (writer, &buflen);
+   req->kmip.data = malloc (buflen);
+   memcpy (req->kmip.data, buf, buflen);
+   req->kmip.len = (uint32_t) buflen;
+}
+
+kms_request_t *
+kms_kmip_request_register_secretdata_new (void *reserved,
+                                          uint8_t *data,
+                                          uint32_t len)
+{
+   /*
+   Create a KMIP Register request with a 96 byte SecretData of this form:
+
+   <RequestMessage tag="0x420078" type="Structure">
+    <RequestHeader tag="0x420077" type="Structure">
+     <ProtocolVersion tag="0x420069" type="Structure">
+      <ProtocolVersionMajor tag="0x42006a" type="Integer" value="1"/>
+      <ProtocolVersionMinor tag="0x42006b" type="Integer" value="4"/>
+     </ProtocolVersion>
+     <BatchCount tag="0x42000d" type="Integer" value="1"/>
+    </RequestHeader>
+    <BatchItem tag="0x42000f" type="Structure">
+     <Operation tag="0x42005c" type="Enumeration" value="3"/>
+     <RequestPayload tag="0x420079" type="Structure">
+      <ObjectType tag="0x420057" type="Enumeration" value="7"/>
+      <TemplateAttribute tag="0x420091" type="Structure">
+      </TemplateAttribute>
+      <SecretData tag="0x420085" type="Structure">
+       <SecretDataType tag="0x420086" type="Enumeration" value="2"/>
+       <KeyBlock tag="0x420040" type="Structure">
+        <KeyFormatType tag="0x420042" type="Enumeration" value="1"/>
+        <KeyValue tag="0x420045" type="Structure">
+         <KeyMaterial tag="0x420043" type="ByteString" value="..."/>
+        </KeyValue>
+       </KeyBlock>
+      </SecretData>
+     </RequestPayload>
+    </BatchItem>
+   </RequestMessage>
+   */
+
+   kmip_writer_t *writer;
+   kms_request_t *req;
+
+   req = calloc (1, sizeof (kms_request_t));
+   req->provider = KMS_REQUEST_PROVIDER_KMIP;
+
+   if (len != 96) {
+      KMS_ERROR (req, "expected SecretData length of 96, got %" PRIu32, len);
+      return req;
+   }
```

Comment:
```suggestion
   if (len != KMS_KMIP_REQUEST_SECRETDATA_LENGTH) {
      KMS_ERROR (req,
                 "expected SecretData length of %" PRIu32 ", got %" PRIu32,
                 (uint32_t) KMS_KMIP_REQUEST_SECRETDATA_LENGTH,
                 len);
      return req;
   }
```
Given this value is constant, I suggest defining this constant and using it to replace/implement `SECRET_DATA_INVALID_LENGTH` in unit tests.

Reply:
Done.